transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,260 @@
+            try {
+                handler.channelAcquired(ch, key);
+                promise.setSuccess(ch);
+            } catch (Throwable cause) {
+                ch.close();
+                promise.setFailure(cause);
+            }
+        } else {
+            ch.close();
+            acquire(key, promise);
+        }
+    }
+
+    private void newChannel(
+            final K key, final Promise<C> promise) {
+        Bootstrap bs = bootstrap.clone(loop(key));
+        ChannelFuture f = bs.attr(KEY, key).connect(key.remoteAddress());
+        if (f.isDone()) {
+            notifyConnect(f, key, promise);
+        } else {
+            f.addListener(new ChannelFutureListener() {
+                @SuppressWarnings("unchecked")
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    notifyConnect(future, key, promise);
+                }
+            });
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void notifyConnect(ChannelFuture future, ChannelPoolKey key, Promise<C> promise) {
+        if (future.isSuccess()) {
+            C ch = (C) future.channel();
+            ch.attr(KEY).set(key);
+            promise.setSuccess(ch);
+        } else {
+            promise.setFailure(future.cause());
+        }
+    }
+
+    @Override
+    public final Future<Boolean> release(C channel) {
+        return release(channel, channel.eventLoop().<Boolean>newPromise());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Future<Boolean> release(final C channel, final Promise<Boolean> promise) {
+        checkNotNull(channel, "channel");
+        checkNotNull(promise, "promise");
+        try {
+            final K key = (K) channel.attr(KEY).getAndSet(null);
+            if (key != null) {
+                ChannelPoolSegment<C> channels = pool.get(key);
+                if (channels == null) {
+                    channels = segmentFactory.newSegment();
+                    ChannelPoolSegment<C> old = pool.putIfAbsent(key, channels);
+                    if (old != null) {
+                        channels = old;
+                    }
+                }
+                final ChannelPoolSegment<C> channelQueue = channels;
+
+                EventLoop loop = channel.eventLoop();
+                if (loop.inEventLoop()) {
+                    doReleaseChannel(channelQueue, key, channel, promise);
+                } else {
+                    loop.execute(new OneTimeTask() {
+                        @Override
+                        public void run() {
+                            doReleaseChannel(channelQueue, key, channel, promise);
+                        }
+                    });
+                }
+            } else {
+                promise.setSuccess(Boolean.FALSE);
+            }
+
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+            channel.close();
```

Comment:
Does it matter if we close before we set the promise?  Should we at least be consistent?